lib/active_model/serializer/adapter/json.rb

```diff
@@ -1,47 +1,128 @@
+    end
+
+    # TODO: remove usage of result instance variable from other flows / tests
+    @result = serialized_hash
+
+    { root => serialized_hash }
+  end
+
+  def fragment_cache(cached_hash, non_cached_hash)
+    ActiveModel::Serializer::Adapter::Json::FragmentCache.new().fragment_cache(cached_hash, non_cached_hash)
+  end
+
+  private
+
+  # iterate through the associations on the serializer,
+  # adding them to the parent as needed (as singular or plural)
+  #
+  # nested_associations is a list of symbols that governs what
+  # associations on the passed in seralizer to include
+  def add_resource_relationships(parent, serializer, nested_associations = {})
+    # have the include array normalized
+    nested_associations = ActiveModel::Serializer::Utils.include_options_to_hash(nested_associations)
+
+    included_associations = if nested_associations.present?
+                              serializer.associations.select { |association|
+                                # nested_associations is a hash of:
+                                #   key => nested association to include
+                                nested_associations.key?(association.name)
+                              }
+                            else
+                              serializer.associations
+                            end
+
+    included_associations.each do |association|
+      serializer = association.serializer
+      opts = association.options
+      key = association.key
+
+      # sanity check if the association has nesting data
+      has_nesting = nested_associations[key].present?
+      if has_nesting
+        include_options_from_parent = { include: nested_associations[key] }
+        opts = opts.merge(include_options_from_parent)
+      end
+
+      if serializer.respond_to?(:each)
+        parent[key] = add_relationships(serializer, opts)
+      else
+        parent[key] = add_relationship(serializer, opts)
+      end
+    end
+
+    parent
+  end
+
+  # add a singular relationship
+  # the options should always belong to the serializer
+  def add_relationship(serializer, options)
+    serialized_relationship = serialized_or_virtual_of(serializer, options)
+
+    nested_associations_to_include = options[:include]
+    if nested_associations_to_include.present?
+      serialized_relationship = add_resource_relationships(
+        serialized_relationship,
+        serializer,
+        nested_associations_to_include)
+    end
+
+    serialized_relationship
+  end
+
+  # add a many relationship
+  def add_relationships(serializer, options)
+    serialize_array(serializer) do |serialized_item, item_serializer|
+      nested_associations_to_include = options[:include]
+
+      if nested_associations_to_include.present?
+        serialized_item = add_resource_relationships(
+          serialized_item,
+          item_serializer,
+          nested_associations_to_include)
+      end
+
+      serialized_item
+    end
+  end
```

Comment:
@joaomdmoura I'd love to see the serializer refactor in a preceding PR to keep the PR-specific stuff separate... but that would need @NullVoxPopuli and your buy-in to keep the process running